File: compiler/src/typecheck/instantiate.rs
use crate::{
    database::{Db, Fact, HiddenNode, NodeRef, Render},
    typecheck::{Replacements, Substitutions, Type, Typed},
};

#[derive(Debug, Clone)]
pub struct Instantiated {
    pub definition: NodeRef,
    pub from: NodeRef,
    pub source_node: NodeRef,
}

impl Fact for Instantiated {}

impl Render for Instantiated {}

pub struct InstantiateContext<'db> {
    pub db: &'db mut Db,
    pub definition: NodeRef,
    pub source_node: NodeRef,
    pub replacements: Replacements,
    pub substitutions: Substitutions,
}

impl InstantiateContext<'_> {
    pub fn instantiate_node(&mut self, node: &NodeRef) -> NodeRef {
        self.replacements.get_or_insert_with(node, || {
            let span = self.db.span(node);
            let instantiated = self.db.node(span, HiddenNode(None));

            self.db.insert(&instantiated, Typed::default());

            self.db.insert(
                &instantiated,
                Instantiated {
                    definition: self.definition.clone(),
                    from: node.clone(),
                    source_node: self.source_node.clone(),
                },
            );

            self.db.graph.replace(node, &instantiated);

            instantiated
        })
    }

    pub fn instantiate_type(&mut self, ty: &Type) -> Type {
        ty.traverse(|ty| match ty {
            Type::Node(node) => Type::Node(self.instantiate_node(&node)),
            Type::Constructed(ty) => {
                let Some(parameter) = &ty.instantiate else {
                    return Type::Constructed(ty);
                };

                self.substitutions.get_or_insert_with(parameter, || {
                    let span = self.db.span(parameter);
                    let substitution = self.db.node(span, HiddenNode(None));

                    self.db.insert(&substitution, Typed::default());

                    self.db.insert(
                        &substitution,
                        Instantiated {
                            definition: self.definition.clone(),
                            from: parameter.clone(),
                            source_node: self.source_node.clone(),
                        },
                    );

                    Type::Node(substitution)
                })
            }
        })
    }

    pub fn instantiate_substitutions(&mut self, substitutions: &Substitutions) -> Substitutions {
        substitutions
            .entries()
            .map(|(parameter, substitution)| (parameter, self.instantiate_type(&substitution)))
            .collect()
    }
}

File: compiler/src/typecheck/Instantiate.swift
public struct Instantiated: Fact {
    public let definition: Node
    public let from: Node
    public let sourceNode: Node

    public func render(into context: RenderContext) {}
}

struct InstantiateContext {
    let db: DB
    let definition: Node
    let sourceNode: Node
    let replacements: Replacements
    let substitutions: Substitutions

    func instantiate(_ node: Node) -> Node {
        if let replacement = self.replacements[node] { return replacement }

        let replacement = self.db.node(isHidden: true)
        self.replacements[node] = replacement

        self.db[replacement, Typed.self] = Typed()

        self.db[replacement, Instantiated.self] = Instantiated(
            definition: self.definition,
            from: node,
            sourceNode: self.sourceNode,
        )

        return replacement
    }

    func instantiate(_ type: Type) -> Type {
        type.traverse { type in
            switch type {
            case .node(let node): return .node(self.instantiate(node))
            case .constructed(let type):
                guard case .parameter(let parameter) = type.tag else { return .constructed(type) }

                if let substitution = self.substitutions[parameter] { return substitution }

                let substitution = self.db.node(isHidden: true)
                self.substitutions[parameter] = .node(substitution)

                self.db[substitution, Typed.self] = Typed()

                self.db[substitution, Instantiated.self] = Instantiated(
                    definition: self.definition,
                    from: parameter,
                    sourceNode: self.sourceNode,
                )

                return .node(substitution)
            }
        }
    }

    func instantiate(_ substitutions: Substitutions) -> Substitutions {
        let newSubstitutions = Substitutions()
        for (parameter, substitution) in substitutions {
            newSubstitutions[parameter] = self.instantiate(substitution)
        }

        return newSubstitutions
    }
}
